### PR TITLE
Fix GCD codec

### LINKED ISSUE #62629: Server crashes after insert into a table with GCD codec (v24.4.1.497)
(win10, wsl, v24.4.1.497)
this request is almost guaranteed to crash my server, somewhere starting from 5 million

CREATE TABLE test_gcd
(
    `test_col` UInt32 CODEC(GCD, LZ4)
)
ENGINE = MergeTree
ORDER BY tuple()
SETTINGS index_granularity = 8192, index_granularity_bytes = 1024

INSERT INTO test_gcd SELECT floor(randUniform(1, 3)) FROM numbers(15000000)
OPTIMIZE TABLE test_gcd FINAL

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed bug in GCD codec implementation that may lead to server crashes.

Fixes: #62629

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 252a82dc387577cf8662d2e53ebe3f9c0282869f with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62853/252a82dc387577cf8662d2e53ebe3f9c0282869f/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by evillique — APPROVED** [review]

File: tests/queries/0_stateless/03093_bug_gcd_codec.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
+-- Tags: long
+
+CREATE TABLE test_gcd(test_col UInt32 CODEC(GCD, LZ4))
+ENGINE = MergeTree
+ORDER BY tuple()
+SETTINGS index_granularity = 8192, index_granularity_bytes = 1024;
+
+INSERT INTO test_gcd SELECT floor(randUniform(1, 3)) FROM numbers(15000000);
```
Comment (evillique) [inline tests/queries/0_stateless/03093_bug_gcd_codec.sql:8]: Does it reproduce on less data to make the test less heavy?

Comment (nickitat) [reply tests/queries/0_stateless/03093_bug_gcd_codec.sql]: yes

**Comment by clickgapai** [comment]
Hi @nickitat @evillique — while reviewing this PR I found the following:

- **Coverage gap:** GCD codec real compression path (gcd_multiplier > 1) untested for any data type — https://github.com/ClickHouse/ClickHouse/pull/104099

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**